### ISSUE TITLE
trainer.save on distributed case

### ISSUE THREAD

**Issue by TLB-MISS** [lucidrains/imagen-pytorch#309]
Hi!
Thank you for always providing good code.

When I use below code, process gets stuck.
```
if not (i % 2000) and trainer.is_main: # is_main makes sure this can run in distributed
        images = trainer.sample(batch_size = 1, stop_at_unet_number = 1, return_pil_images = True) # returns List[Image]
        images[0].save(f'./results/sample_step{i}.png')
        trainer.save(f'./checkpoints/checkpoint_step{i}.pt') // error on this line
```
The error is same with [this](https://github.com/lucidrains/imagen-pytorch/issues/308)

I think it is caused by this [line](https://github.com/lucidrains/imagen-pytorch/blob/e077796fc13fab751bcaa33565b46ccbb56e58a7/imagen_pytorch/trainer.py#L675).
All process cannot reach the above line except for the main trainer's.
So I delete the above line and try the following code. 
```
if not (i % 2000): # is_main makes sure this can run in distributed
        trainer.accelerator.wait_for_everyone()
        if trainer.is_main:
            images = trainer.sample(batch_size = 1, stop_at_unet_number = 1, return_pil_images = True) # returns List[Image]
            images[0].save(f'./results/sample_step{i}.png')
            trainer.save(f'./checkpoints/checkpoint_step{i}.pt')
```
It seems to work fine.

But is it proper way?
If it is not proper, is there any suggested way to save checkpoint with accelerate(multi-GPU case)?

**Comment by SushantGautam** [comment]
Same problem

**Comment by ChawDoe** [comment]
> Hi! Thank you for always providing good code.
> 
> When I use below code, process gets stuck.
> 
> ```
> if not (i % 2000) and trainer.is_main: # is_main makes sure this can run in distributed
>         images = trainer.sample(batch_size = 1, stop_at_unet_number = 1, return_pil_images = True) # returns List[Image]
>         images[0].save(f'./results/sample_step{i}.png')
>         trainer.save(f'./checkpoints/checkpoint_step{i}.pt') // error on this line
> ```
> 
> The error is same with [this](https://github.com/lucidrains/imagen-pytorch/issues/308)
> 
> I think it is caused by this [line](https://github.com/lucidrains/imagen-pytorch/blob/e077796fc13fab751bcaa33565b46ccbb56e58a7/imagen_pytorch/trainer.py#L675). All process cannot reach the above line except for the main trainer's. So I delete the above line and try the following code.
> 
> ```
> if not (i % 2000): # is_main makes sure this can run in distributed
>         trainer.accelerator.wait_for_everyone()
>         if trainer.is_main:
>             images = trainer.sample(batch_size = 1, stop_at_unet_number = 1, return_pil_images = True) # returns List[Image]
>             images[0].save(f'./results/sample_step{i}.png')
>             trainer.save(f'./checkpoints/checkpoint_step{i}.pt')
> ```
> 
> It seems to work fine.
> 
> But is it proper way? If it is not proper, is there any suggested way to save checkpoint with accelerate(multi-GPU case)?

The tricky thing is that the code you write is a bug. You don't need to write is_main before trainer.save(). Because trainer.save() will automatically wait every one. If you write is_main, it could not wait every one so it gets stuck. The right code for saving checkpoints is: 
`` 
if i % 2000 == 0:
    trainer.save(f'./checkpoints/checkpoint_{i}.pt')
``

**Comment by santisy** [comment]
Yes. Notable findings @ChawDoe . Also for distributed training, please make sure that you are loading different data at every spawned process, namely paying attention to your random seed setting and set it related to your rank number at current process.
